### ISSUE TITLE
Group or Workspace

### ISSUE THREAD

**Issue by jackycute** [jackycute/HackMD#12]
Need a group page or a workspace for a small team.
Should use subdomain to provided better experience.

**Comment by janniklorenz** [comment]
What about defining workspaces as "a place where notes are" with following attributes:
- name, which it will be accessible under (/public, /teama)
- Read access (user, group, world)
- Write access (user, group, world)

This requires a clean user database, so it's maybe the best to go away from the passport module(s) and directly implement the user/group register feature. 

We could create automatically a workspace for every user and group so they directly have there own private place. Additional workspaces can be defined by users, who can edit them.

**Comment by janniklorenz** [comment]
But maybe we should separate the whole user/ login section from the core feature, and require the core which could be an express instance.

What do you think?

**Comment by jackycute** [comment]
You mean to build a module to run our own instance, right?
I think that's a good solution.

I agreed that we should build our own group/workspace register because it's very unique.
But in user signin part, I'm pretty satisfied with passport because that part can be really hard to do right and have very high complexity.

group/workspace should only for multi users situation, if it's personal using, history is enough in my opinion.
And what about sub domain? It's like `yourdomain.hackmd.io` which I think pretty sweet for teams and more straightforward to understand.

**Comment by janniklorenz** [comment]
# Split
Yes I mean to build a node module which contains its own express instance and all necessary stuff for the editor fronted, like socket.io, html, css and the client side javascript. (Editor Module)

So that we can use it in the main module (User Module) by simply binding the editor with an `app.use("/note123", oureModule);` to the express instance. 
The editor module should provide a simple api to manage the data in the user module. 

This would simplify the maintenance form both of them, and gives us the opportunity to use the realtime editor somewhere else, for example a simple page to create notes without login. 

# Login
I'm not familiar with the passport module, but if we can combine that with a local user database, why not. For me, a simple username/ password login with an register form would be enough. 

# Subdomain
I don't know if a small team want to host this stuff somewhere else. Hosting a node application isn't expensive and complicated any more. And I think the domain has noting to do with the the application running behind it.

**Comment by jackycute** [comment]
I think you're right.
But it's somehow will go different route from current position.
I will figure these out in this weekend.
Anyway, I think we need more people involved to do this, what do you think?
Thank you so much :smile:

**Comment by janniklorenz** [comment]
Yes, but to get more people involved we have to provide a good documentation from the core functions, to make it easier to get in to it.

Did you write the complete code, or just the server side?

**Comment by jackycute** [comment]
I wrote both client side and server side.
But using lots of packages of other open source things.

I decided to split whole HackMD into modules.
And there will be an organization, but unfortunately /hackmd is already taken.
So maybe new repo will just in my old team organization.
Will let you know when it gets done.

And everything should be documented, I just not having enough time to do it.
But it can be done with new rewrite codes.
Thank you.

**Comment by eshao** [comment]
Yes, I think this is the only core feature that would make hackmd much better than all the other alternatives. :)

**Comment by ToniIltanen** [comment]
Any progress info on this one?

**Comment by jackycute** [comment]
Not now, maybe after user space feature #166 done will turn to this.

**Comment by grexe** [comment]
I was excited to try out HackMD for our team, but sadly the lack of a shared space where you automatically see all notes as covered in #322 and also here is holding back adoption:(

As it is now, you can only view your own notes from your history or use a shared link, so it's only useful for collaborating on specific documents, but not usable as a Wiki.

I really hope this feature gets prioritized as I'd really like to stick with HackMD and avoid using a traditional Wiki.

**Comment by grexe** [comment]
@jackycute any chance team support is getting implemented in the near future? Would love to try collaborative score editing with the recent addition of abc.js support;=)

**Comment by jackycute** [comment]
Hey @grexe 
Sorry for late reply.
I could only say we might catch this up in few months.
There are lots of dependencies before make this feature happen.
Thanks for continuing reaching us, we'll do our best 😄

**Comment by grexe** [comment]
thanks for the update @jackycute - no problem, I underestimated the work involved. Sadly this will be too late for us, but I'll keep a close eye on this exciting project!

**Comment by evanslify** [comment]
Can we see this feature come to light in the near future?
